tests/federation_device_list_update_test.go

```diff
@@ -0,0 +1,165 @@
+	deployment := complement.Deploy(t, 2)
+	defer deployment.Destroy(t)
+
+	syncHasDeviceListChange := func(changed []string, left []string) client.SyncCheckOpt {
+		sort.Strings(changed)
+		sort.Strings(left)
+		return func(clientUserID string, topLevelSyncJSON gjson.Result) error {
+			dl := topLevelSyncJSON.Get("device_lists")
+			changedJSON := dl.Get("changed").Array()
+			leftJSON := dl.Get("left").Array()
+			gotChanged := make([]string, len(changedJSON))
+			gotLeft := make([]string, len(leftJSON))
+			for i := range gotChanged {
+				gotChanged[i] = changedJSON[i].Str
+			}
+			for i := range gotLeft {
+				gotLeft[i] = leftJSON[i].Str
+			}
+			sort.Strings(gotChanged)
+			sort.Strings(gotLeft)
+			var changedMatch, leftMatch bool
+			if reflect.DeepEqual(changed, gotChanged) {
+				changedMatch = true
+			}
+			if reflect.DeepEqual(left, gotLeft) {
+				leftMatch = true
+			}
+			if changedMatch && leftMatch {
+				return nil
+			}
+			return fmt.Errorf("syncHasDeviceListChange: got changed %v want %v, got left %v want %v", gotChanged, changed, gotLeft, left)
+		}
+	}
+
+	testCases := []struct {
+		name            string
+		makeUnreachable func(t *testing.T)
+		makeReachable   func(t *testing.T)
+	}{
+		{
+			name:            "good connectivity",
+			makeUnreachable: func(t *testing.T) {},
+			makeReachable:   func(t *testing.T) {},
+		},
+		{
+			// cut networking but keep in-memory state
+			name: "interrupted connectivity",
+			makeUnreachable: func(t *testing.T) {
+				deployment.StopServer(t, "hs2")
+			},
+			makeReachable: func(t *testing.T) {
+				deployment.StartServer(t, "hs2")
+			},
+		},
+		{
+			// interesting because this nukes memory
+			name: "stopped server",
+			makeUnreachable: func(t *testing.T) {
+				deployment.StopServer(t, "hs2")
+			},
+			makeReachable: func(t *testing.T) {
+				// kick over the sending server first to see if the server
+				// remembers to resend on startup
+				deployment.StopServer(t, "hs1")
+				deployment.StartServer(t, "hs1")
+				// now make the receiving server reachable.
+				deployment.StartServer(t, "hs2")
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			alice := deployment.Register(t, "hs1", helpers.RegistrationOpts{
+				LocalpartSuffix: "alice",
+				Password:        "this is alices password",
+			})
+			bob := deployment.Register(t, "hs2", helpers.RegistrationOpts{
+				LocalpartSuffix: "bob",
+			})
+			// they must share a room to get device list updates
+			roomID := alice.MustCreateRoom(t, map[string]interface{}{
+				"preset": "private_chat",
+				"invite": []string{bob.UserID},
+				// we strictly don't need to make this an encrypted room, but there are some
+				// issues which state that we should only share device list updates for
+				// users in shared encrypted rooms, so let's ensure we do that.
+				// See https://github.com/matrix-org/synapse/issues/7524
```

Comment:
Yea, I wish we didn't send all device list changes over federation if not necessary. (Both, Dendrite and Synapse do, afaik)